src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesCcm.Unix.cs

```diff
@@ -0,0 +1,211 @@
+                    // this shouldn't happen
+                    throw new CryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesCcmGetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+            }
+        }
+
+        public void Decrypt(ReadOnlySpan<byte> nonce, ReadOnlySpan<byte> ciphertext, ReadOnlySpan<byte> tag, Span<byte> plaintext, ReadOnlySpan<byte> associatedData)
+        {
+            AesAEAD.CheckArguments(nonce, plaintext, ciphertext, tag);
+            CheckParameters(nonce.Length, tag.Length);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+
+            int keyLength = _keyHandle.Key.Length * 8;
+            using (SafeEvpCipherCtxHandle ctx = Interop.Crypto.EvpCipherCreatePartial(
+                    GetCipher(keyLength),
+                    keyLength,
+                    0))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(ctx);
+
+                if (!Interop.Crypto.EvpAesCcmSetNonceLength(ctx, nonce.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesCcmSetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherSetKeyAndIV(ctx, ref MemoryMarshal.GetReference(_keyHandle.Key), ref MemoryMarshal.GetReference(nonce), 0 /*decrypting*/))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (associatedData.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref nullRef, ciphertext.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref MemoryMarshal.GetReference(associatedData), associatedData.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+                }
+
+                int plaintextBytesWritten;
+                if (ciphertext.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref MemoryMarshal.GetReference(plaintext), out plaintextBytesWritten, ref MemoryMarshal.GetReference(ciphertext), ciphertext.Length))
+                    {
+                        throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+                    }
+                }
+                else
+                {
+                    byte nonEmptyArray = 123;
```

Comment:
I'm not sure I understand what this block is for... is it to handle a 0 length ciphertext that still has associated data? If so, I don't understand why we're passing in this value

Reply:
yes, will add comment